packages/@jsii/kernel/src/disk-cache/disk-cache.ts

```diff
@@ -0,0 +1,244 @@
+      for (const file of directoriesUnder(dir)) {
+        yield* inDirectory(file);
+      }
+    }
+  }
+}
+
+export class Entry {
+  public constructor(public readonly path: string) {}
+
+  public get atime(): Date {
+    try {
+      const stat = statSync(this.markerFile);
+      return stat.atime;
+    } catch (err: any) {
+      if (err.code !== 'ENOENT') {
+        throw err;
+      }
+      return new Date(0);
+    }
+  }
+
+  public get pathExists() {
+    return existsSync(this.path);
+  }
+
+  private get lockFile(): string {
+    return `${this.path}.lock`;
+  }
+
+  private get markerFile(): string {
+    return join(this.path, MARKER_FILE_NAME);
+  }
+
+  public lock<T>(cb: (entry: LockedEntry) => T): T {
+    mkdirSync(dirname(this.path), { recursive: true });
+    lockSync(this.lockFile, { retries: 12, stale: 5_000 });
+    let disposed = false;
+    try {
+      return cb({
+        delete: () => {
+          if (disposed) {
+            throw new Error(
+              `Cannot delete ${this.path} once the lock block as returned!`,
+            );
+          }
+          rmSync(this.path, { force: true, recursive: true });
+        },
+        write: (name, content) => {
+          if (disposed) {
+            throw new Error(
+              `Cannot write ${join(
+                this.path,
+                name,
+              )} once the lock block as returned!`,
+            );
+          }
+
+          mkdirSync(dirname(join(this.path, name)), { recursive: true });
+          writeFileSync(join(this.path, name), content);
+        },
+        touch: () => {
+          if (disposed) {
+            throw new Error(
+              `Cannot touch ${this.path} once the lock block as returned!`,
+            );
+          }
+          if (this.pathExists) {
+            if (existsSync(this.markerFile)) {
+              const now = new Date();
+              utimesSync(this.markerFile, now, now);
+            } else {
+              writeFileSync(this.markerFile, '');
+            }
+          }
+        },
+      });
+    } finally {
+      disposed = true;
+      unlockSync(this.lockFile);
+    }
+  }
+
+  public read(file: string): Buffer | undefined {
+    try {
+      return readFileSync(join(this.path, file));
+    } catch (error: any) {
+      if (error.code === 'ENOENT') {
+        return undefined;
+      }
+      throw error;
+    }
+  }
+}
+
+export interface LockedEntry {
+  delete(): void;
+  write(name: string, data: Buffer): void;
+
+  touch(): void;
+}
+
+function* directoriesUnder(
+  root: string,
+  recursive = false,
+  ignoreErrors = true,
+): Generator<string, void, void> {
+  for (const file of readdirSync(root)) {
+    const path = join(root, file);
+    try {
+      const stat = statSync(path);
+      if (stat.isDirectory()) {
+        if (recursive) {
+          yield* directoriesUnder(path, recursive, ignoreErrors);
+        }
+        yield path;
+      }
+    } catch (error) {
+      if (!ignoreErrors) {
```

Comment:
Is there a scenario in which the errors should both be captured (for investigation) and ignored?

Reply:
Maybe... I decided for this to be as silent as possible, to reduce the chances it causes something to fail that wouldn't have if the cache wasn't enabled.